WordPress/wordpress-develop
# Add block-based widgets editor

Trac ticket: https://core.trac.wordpress.org/ticket/51506

Supersedes https://github.com/WordPress/wordpress-develop/pull/603.
Requires https://github.com/WordPress/wordpress-develop/pull/1293.

This moves the the block-based widget editors over from Gutenberg into Core.

Broadly speaking, there's four parts to this:

1. `@wordpress/edit-widgets`, `@wordpress/customize-widgets`, and `@wordpress/widgets` are added as dependencies. These packages contain the block editors.

2. Supporting infrastructure such as `WP_Widget_Block` has been copied over from Gutenberg.

3. `wp-admin/widgets.php` has been modified to branch between the old form-based editor `wp-admin/widgets-form.php` and the new block-based editor `wp-admin/widgets-form-blocks.php` depending on `get_theme_support( 'widgets-block-editor' )` and `use_widgets_block_editor`.

4. `WP_Customize_Widgets` has been modified to use `WP_Sidebar_Block_Editor_Control` which contains the block editor instead of `WP_Widget_Area_Customize_Control` depending on `get_theme_support( 'widgets-block-editor' )` and `use_widgets_block_editor`.

This PR contains `WP_REST_Sidebars_Controller` and `WP_REST_Widget_Utils_Controller` so that it's easier to test, but these parts should be committed separately. @TimothyBJacobs is working on it.

#### How to test

1. `npm run env:start`
2. `npm run env:install`
3. `npm run dev`
3. Go to Appearance → Widgets to see the `widgets.php` block editor.
4. Go to Appearance → Customize → widgets to see the `customize.php` block editor.

### Tasks remaining

- [x] REST API changes (https://github.com/WordPress/wordpress-develop/pull/1293) need to be committed separately. 
- [x] Customizer changes are not being saved.
- [x] Figure out if any of `WP_Customize_Widgets::enqueue_scripts` can be skipped if block editor is enabled.
- [x] `should_load_block_editor_scripts_and_styles` filter is not run in Customizer.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**
